src/Symfony/Installer/SelfUpdateCommand.php

```diff
@@ -45,71 +55,106 @@ public function isEnabled()
         }
 
         if ($localVersion === $remoteVersion) {
-            $output->writeln('<info>Symfony Installer is already up to date.</info>');
-
-            return 0;
+            $this->output->writeln('<info>Symfony Installer is already up to date.</info>');
+            $isUpdated = true;
         }
 
-        $remoteFilename = 'http://symfony.com/installer';
-        $localFilename = realpath($_SERVER['argv'][0]) ?: $_SERVER['argv'][0];
-        $tempDir = is_writable(dirname($localFilename)) ? dirname($localFilename) : sys_get_temp_dir();
+        return $isUpdated;
+    }
 
+    private function downloadNewVersion()
+    {
         // check for permissions in local filesystem before start downloading files
-        if (!is_writable($localFilename)) {
-            throw new \RuntimeException('Symfony Installer update failed: the "'.$localFilename.'" file could not be written');
+        if (!is_writable($this->localInstallerFile)) {
+            throw new \RuntimeException('Symfony Installer update failed: the "'.$this->localInstallerFile.'" file could not be written');
         }
-        if (!is_writable($tempDir)) {
-            throw new \RuntimeException('Symfony Installer update failed: the "'.$tempDir.'" directory used to download the temporary file could not be written');
+
+        if (!is_writable($this->tempDir)) {
+            throw new \RuntimeException('Symfony Installer update failed: the "'.$this->tempDir.'" directory used to download files temporarily could not be written');
+        }
+
+        if (false === $newInstaller = @file_get_contents($this->remoteInstallerFile)) {
+            throw new \RuntimeException('The new version of the Symfony Installer couldn\'t be downloaded from the server.');
         }
 
-        if (false === @file_get_contents($remoteFilename)) {
-            $output->writeln('<error>The new version of the Symfony Installer couldn\'t be downloaded from the server.</error>');
+        $this->fs->dumpFile($this->tempInstallerFilePath, $newInstaller, 0777 & ~umask());
+
+        return $this;
+    }
 
-            return 1;
+    private function checkNewVersionIsValid()
+    {
+        // creating a Phar instance for an existing file is not allowed
+        // when the Phar extension is in readonly mode
+        if (!ini_get('phar.readonly')) {
+            // test the phar validity
+            $phar = new \Phar($this->tempInstallerFilePath);
+
+            // free the variable to unlock the file
+            unset($phar);
         }
 
-        try {
-            $tempFilename = $tempDir.'/'.basename($localFilename, '.phar').'-temp.phar';
+        return $this;
+    }
 
-            $fs->copy($remoteFilename, $tempFilename);
-            $fs->chmod($tempFilename, 0777 & ~umask());
+    private function backupOldVersion()
+    {
+        $this->fs->rename($this->localInstallerFile, $this->previousInstallerFile);
```

Comment:
this is what composer is doing (except it does not even remove it later because it allows ``composer self-update --rollback``)